### PR TITLE
Introduce GetTypeByMetadataName(KnownType)

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ForeachLoopExplicitConversionCodeFix.cs [hunk 1]
```diff
@@ -67,15 +68,13 @@ protected override Task RegisterCodeFixesAsync(SyntaxNode root, CodeFixContext c
...
-        private const string ofTypeExtensionClass = "System.Linq.Enumerable";
-
         private static SyntaxNode CalculateNewRoot(SyntaxNode root, ForEachStatementSyntax foreachSyntax, SemanticModel semanticModel)
         {
             var collection = foreachSyntax.Expression;
             var typeName = foreachSyntax.Type.ToString();
             var invocationToAdd = GetOfTypeInvocation(typeName, collection);
             var namedTypes = semanticModel.LookupNamespacesAndTypes(foreachSyntax.SpanStart).OfType<INamedTypeSymbol>();
-            var isUsingAlreadyThere = namedTypes.Any(nt => nt.ToDisplayString() == ofTypeExtensionClass);
+            var isUsingAlreadyThere = namedTypes.Any(KnownType.System_Linq_Enumerable.Matches);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ForeachLoopExplicitConversionCodeFix.cs:77]: So much better 😄

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterMissingOptionalCodeFix.cs [hunk 2]
```diff
@@ -60,26 +61,22 @@ protected override async Task RegisterCodeFixesAsync(SyntaxNode root, CodeFixCon
             context.RegisterCodeFix(
                 CodeAction.Create(
                     Title,
-                    c =>
+                    _ =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterMissingOptionalCodeFix.cs:64]: This is not a discard. It is a variable with a less clear name `_` that can be used to do `_.ThrowIfCancellationRequested();`. I prefer the old name.
```suggestion
                    c =>
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/OptionalParameterWithDefaultValueCodeFix.cs [hunk 3]
```diff
@@ -60,10 +61,9 @@ protected override async Task RegisterCodeFixesAsync(SyntaxNode root, CodeFixCon
             context.RegisterCodeFix(
                 CodeAction.Create(
                     Title,
-                    c =>
+                    _ =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/OptionalParameterWithDefaultValueCodeFix.cs:64]: Same
```suggestion
                    c =>
```

File: analyzers/src/SonarAnalyzer.Common/Common/CompilationExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Common;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Common/CompilationExtensions.cs:24]: Hiding this one in yet another namespace (some of them are in `Extensions`, some of them are in `Helpers`) would make it hard to use.

There's a bigger problem with assembly separation and their namespaces. But for now, let's play with what we have.
```suggestion
namespace SonarAnalyzer.Extensions;
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Let's align the namespace with the other extensions

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterMissingOptionalCodeFix.cs (hunk 2 above)
Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterMissingOptionalCodeFix.cs:64]: It tells that the variable is not used. It's a common used pattern.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterMissingOptionalCodeFix.cs:64]: I've seen that pattern required on method parameters too by the community. And it was not a good idea to have public `Method(int _)`

Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterMissingOptionalCodeFix.cs:64]: I don't understand why. It makes it obvious that the parameter is not used and points out that maybe a refactoring is needed.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterMissingOptionalCodeFix.cs:64]: We can't change the source of the signature anyway. None of the variant (`_`, `c`) is used in the code. The difference is that we loose information about what is available.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**